Allow application/x-www-form-urlencoded request body payloads

sidraval opened the pull request:
The API is basically:

``` swift
urlRequest.formStylePayload = [
  "firstKey": "firstValue",
  "secondKey": "secondValue"
]
```

It will automatically percent-encode characters that are forbidden in the URL (e.g. if you wanted to send the value `value=&%`), and then it will base64 encode the string.

WRT #63

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/thoughtbot/swish/64)

<!-- Reviewable:end -->

Commit:
Allow form style request body payloads

sidraval commented:
Closes #63

gfontenot commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+public extension NSURLQueryItem {
+  static func initAndEncode(key: String, value: String) -> NSURLQueryItem? {
```
It seems weird that we need this. Does `NSURLQueryItem(name: value:)` not do this automatically?

gfontenot replied:
It _seems_ like we should leave this up to the user. Is that crazy?

gfontenot commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+  private static func encode(s: String) -> String? {
+    let characterSet = NSMutableCharacterSet()
+    let disallowed = NSCharacterSet.URLQueryAllowedCharacterSet()
+    characterSet.formUnionWithCharacterSet(disallowed.invertedSet)
```
What do you think about moving `invertedSet` to the above line so that `disallowed` is named properly from the jump?

gfontenot commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+    let characterSet = NSMutableCharacterSet()
+    let disallowed = NSCharacterSet.URLQueryAllowedCharacterSet()
+    characterSet.formUnionWithCharacterSet(disallowed.invertedSet)
+    characterSet.addCharactersInString("&=")
```
these aren't already included in the query allowed character set?

gfontenot replied:
it's too bad there isn't a pure version of this that returns a new set

gfontenot commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  private static func encode(s: String) -> String? {
+    let characterSet = NSMutableCharacterSet()
```
What about `NSMutableCharacterSet(charactersInString: "&=")` here to remove the mutation on line 31?

gfontenot commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+  }
+}
+
+private extension Array {
```
Can we move this out to its own file and make it internal?

gfontenot commented on Source/Extensions/NSURLRequest.swift:
```diff
@@ -22,4 +34,32 @@ public extension NSMutableURLRequest {
+private extension CollectionType where Generator.Element == NSURLQueryItem {
+  var queryString: String {
+    return sort { $0.name < $1.name }
+    .reduce([]) { accum, el in
```
I feel like this should be indented

gfontenot commented:
I have some questions that expose my general lack of familiarity with `NSURLQueryComponents`, but I _think_ the rest of this makes sense.

tonyd256 commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+    return result ?? []
+  }
+
+  private static func stringToQueryItem(accum: [NSURLQueryItem]?, el: [String]) -> [NSURLQueryItem]? {
```
does this have to take optional `accum` and return an optional? I think this function could remove those and be fine.

tonyd256 commented on Source/Extensions/NSURLRequest.swift:
```diff
@@ -10,6 +10,18 @@ public extension NSURLRequest {
     return jsonArray ?? [:]
   }
+
+  var formStylePayload: [String: String]? {
```
`formStylePayload` doesn't seem right ... if we look at the encoding types, you're dealing with `application/x-www-form-urlencoded` right? so maybe `urlEncodedPayload`?

tonyd256 replied:
Does Swish set the `Content-Type` header? Can't find it in the repo

gfontenot replied:
No, that's up to each individual request and is a consumer level concern.

dazmuda commented on Source/Extensions/NSURLQueryItem.swift:
```diff
@@ -0,0 +1,41 @@
+    let result = String(data: data, encoding: NSUTF8StringEncoding)
+      .flatMap { $0.componentsSeparatedByString("&") }?
+      .map { $0.componentsSeparatedByString("=") }
+      .reduce([], combine: stringToQueryItem)
```
I think this could just be a map, no?
You're taking an array of strings and turning them into an array of NSURLQueryItems.

dazmuda replied:
Also is there maybe a way to use NSURLComponents to split up this URL rather than treating it as a string? Like NSURLComponents(result).queryItems or some such?

sharplet committed:
Update Quick + Nimble

sharplet committed:
Rename formStylePayload to urlEncodedPayload

sharplet commented:
After some investigation, it looks like the current implementation will reject some valid query strings. Specifically, the `=<value>` portion of a parameter is completely optional, meaning the string `novalue` is perfectly valid.

Using `[String: String]`, we're not going to be able to model query strings according to spec. In practice this might be fine, but as a general-purpose networking library this doesn't feel quite right to me.

What about switching to `[URLQueryItem]`? If we did make that change, it would probably make sense to remove the percent-encoding step we're currently performing, and use `URLComponents` to encode the query string. We could potentially expose `URLQueryItem.encode` and `URLQueryItem.decode` for users who want to opt in to the extra level of percent-escaping.

gfontenot commented:
I have no problem with that

sharplet committed:
Model form-urlencoded payloads as [URLQueryItem]

Modelling application/x-www-form-urlencoded payloads as `[String: String]`
is problematic, because it's perfectly valid to have query parameters
with no associated value. That is, `param=` and `param` are two distinct
query paramets: the first with an empty value, and the second with no
value.

This change replaces `[String: String]` with `[URLQueryItem]`, solving
the "no value" problem, and removing the problematic extra layer of URL
encoding. This simpler implementation doesn't make too many extra
guarantees, but still allows users to apply their own extra levels of
encoding/decoding when the manually construct or read instances of
`URLQueryItem`.

sharplet commented:
Ready for re-review 👍

sharplet committed:
Drop NS prefix

tonyd256 commented on Source/Extensions/URLRequest.swift:
```diff
@@ -0,0 +1,27 @@
+    }
+  }
+
+  @available(macOSApplicationExtension 10.10, *)
```
is this not possible on iOS or does this mean to restrict macOS and not restrict iOS?

sharplet replied:
This only introduces a minimum version restriction for macOS application extensions. The `*` means that all other platforms can use it.

tonyd256 commented on Source/Extensions/URLRequest.swift:
```diff
@@ -0,0 +1,27 @@
+    set {
+      var components = URLComponents()
+      components.queryItems = newValue
+      httpBody = components.query!.data(using: .utf8)
```
should this be a `!`? ... can the user potentially send something to this that causes this to crash?

tonyd256 replied:
The JSON version is setting this with an optional so maybe it's ok to do that here too?

sharplet replied:
Unless I'm mistaken, I believe that setting `queryItems` should always update `query`. My assumption is that if you can construct a `URLQueryItem`, you can always get a valid query string. The return type of this expression is still optional, but then maybe it would be fine to just be more conservative and use optional chaining here?

tonyd256 replied:
if you can simply replace the `!` with a `?` and it all still works, then I think that's better.

sharplet committed:
fixup! Model form-urlencoded payloads as [URLQueryItem]
